odoo/addons/base/tests/test_misc.py

```diff
@@ -357,3 +358,95 @@ def test_latin(self):
     def test_non_latin(self):
         self.assertEqual(remove_accents('العربية'), 'العربية')
         self.assertEqual(remove_accents('русский алфавит'), 'русскии алфавит')
+
+class TestAddonsFileAccess(BaseCase):
+
+    def test_file_path(self):
+        def assertCannotAccess(path, ExceptionType=FileNotFoundError, filter_ext=None):
+            with self.assertRaises(ExceptionType):
+                file_path(path, filter_ext=filter_ext)
+
+        # absolute path
+        self.assertEqual(__file__, file_path(__file__))
+        self.assertEqual(__file__, file_path(__file__, filter_ext=None)) # means "no filter" too
+        self.assertEqual(__file__, file_path(__file__, filter_ext=('.py',)))
+
+        # directory target is ok
+        self.assertEqual(os.path.dirname(__file__), file_path(os.path.join(__file__, '..')))
+
+        # relative path
+        relpath = os.path.join(*(__file__.split(os.sep)[-3:])) # 'base/tests/test_misc.py'
+        self.assertEqual(__file__, file_path(relpath))
+        self.assertEqual(__file__, file_path(relpath, filter_ext=('.py',)))
+
+        # leading 'addons/' is ignored if present
+        relpath = os.path.join('addons', relpath) # 'addons/base/tests/test_misc.py'
+        self.assertEqual(__file__, file_path(relpath))
+
+        # files in root_path are allowed
+        self.assertTrue(file_path('tools/misc.py'))
+
+        # errors when outside addons_paths
+        assertCannotAccess('/doesnt/exist')
+        assertCannotAccess('/tmp')
+        assertCannotAccess('../../../../../../../../../tmp')
+        assertCannotAccess(os.path.join(__file__, '../../../../../'))
+
+        # data_dir is forbidden
+        assertCannotAccess(config['data_dir'])
+
+        # errors for illegal extensions
+        assertCannotAccess(__file__, ValueError, filter_ext=('.png',))
+        # file doesnt exist but has wrong extension
+        assertCannotAccess(__file__.replace('.py', '.foo'), ValueError, filter_ext=('.png',))
+
+    def test_file_open(self):
+        # The needle includes UTF8 so we test reading non-ASCII files at the same time.
+        # This depends on the system locale and is harder to unit test, but if you manage to run the
+        # test with a non-UTF8 locale (`LC_ALL=fr_FR.iso8859-1 python3...`) it should not crash ;-)
+        test_needle = "A needle with non-ascii bytes: ♥"
+
+        def assertCanRead(path, needle='', mode='r', filter_ext=None):
+            with file_open(path, mode, filter_ext) as f:
+                self.assertIn(needle, f.read())
+
+        def assertCannotRead(path, ExceptionType=FileNotFoundError, filter_ext=None):
+            with self.assertRaises(ExceptionType):
+                file_open(path, filter_ext=filter_ext)
```

Comment:
Maybe move these to methods or global functions? Especially since `assertCannotRead` is being defined in both test cases.

Reply:
moved to methods, I like to keep the `assertCannotRead()` separate from `assertCannotAccess()` for readability, so there are still 3 of them ;)